relay-server/src/services/buffer/envelope_repository/sqlite.rs

```diff
@@ -0,0 +1,662 @@
+    disk_batch_size: usize,
+    max_disk_size: usize,
+}
+
+impl SqliteEnvelopeRepository {
+    /// Creates a new [`SqliteEnvelopeRepository`] instance.
+    pub async fn new(config: &Config) -> Result<Self, SqliteEnvelopeRepositoryError> {
+        let envelope_store = SqliteEnvelopeStore::prepare(config).await?;
+        Ok(Self {
+            envelope_stacks: HashMap::new(),
+            envelope_store,
+            buffered_envelopes_size: 0,
+            disk_batch_size: config.spool_envelopes_stack_disk_batch_size(),
+            max_disk_size: config.spool_envelopes_max_disk_size(),
+        })
+    }
+
+    /// Creates a new [`SqliteEnvelopeRepository`] instance given a [`SqliteEnvelopeStore`].
+    pub fn new_with_store(config: &Config, envelope_store: SqliteEnvelopeStore) -> Self {
+        Self {
+            envelope_stacks: HashMap::new(),
+            envelope_store,
+            buffered_envelopes_size: 0,
+            disk_batch_size: config.spool_envelopes_stack_disk_batch_size(),
+            max_disk_size: config.spool_envelopes_max_disk_size(),
+        }
+    }
+
+    /// Initializes the envelope repository.
+    ///
+    /// Retrieves the project key pairs from the envelope store and creates
+    /// an initialization state.
+    pub async fn initialize(&mut self) -> InitializationState {
+        match self.envelope_store.project_key_pairs().await {
+            Ok(project_key_pairs) => {
+                self.initialize_empty_stacks(&project_key_pairs);
+                InitializationState::new(project_key_pairs)
+            }
+            Err(error) => {
+                relay_log::error!(
+                    error = &error as &dyn Error,
+                    "failed to initialize the sqlite stack repository"
+                );
+                InitializationState::empty()
+            }
+        }
+    }
+
+    /// Pushes an envelope to the repository for the given project key pair.
+    ///
+    /// If the spool threshold is exceeded, it may trigger spooling to disk.
+    pub async fn push(
+        &mut self,
+        project_key_pair: ProjectKeyPair,
+        envelope: Box<Envelope>,
+    ) -> Result<(), SqliteEnvelopeRepositoryError> {
+        if self.above_spool_threshold() {
+            self.spool_to_disk().await?;
+        }
+
+        self.envelope_stacks
+            .entry(project_key_pair)
+            .or_default()
+            .envelopes
+            .push(envelope);
+
+        self.buffered_envelopes_size += 1;
+
+        Ok(())
+    }
+
+    /// Peeks at the next envelope for the given project key pair without removing it.
+    ///
+    /// If no envelope is in the buffer, it will be loaded from disk and a reference will be
+    /// returned.
+    pub async fn peek(
+        &mut self,
+        project_key_pair: ProjectKeyPair,
+    ) -> Result<Option<&Envelope>, SqliteEnvelopeRepositoryError> {
+        if self.memory_empty(project_key_pair) && self.should_check_disk(project_key_pair) {
```

Comment:
This function does the same lookup a few times, we could instead get the stack object once and then interact with it.

```suggestion
        let Some(stack) = self.envelope_stacks.get(&project_key_pair) else { return Ok(None) };
        if stack.memory_empty() && stack.check_disk {
```

Reply:
It was done this way because we were dealing with the default case in which no data for the project key pair is there, but I feel like it's better to just bail if we don't have that `project_key_pair` in memory, since we assume that if it is not there it means no data about it exists.